fix: deCONZ: ZGP implementation

manup opened the pull request:
The old ZGP frame parser used some magic numbers which often shoot over  the actual received buffer. In the new adapter implementation frames are exposed with their actual size, which as side effect caused the frame parser to have invalid negative offsets.

The new implementation is a port from deCONZ code without fixed magic numbers.

Tested with a old Hue Tap and Friends of Hue switch.

Note that the old implementation likely didn't support "joining" without ZGP proxies since any frame length >= 30 was treated as commissioning command (0x04), which the controller rejected.

Issue: https://github.com/Koenkk/zigbee2mqtt/issues/27888

Nerivec commented on src/adapter/deconz/driver/frameParser.ts:
```diff
@@ -440,41 +441,152 @@ function parseDeviceStateChangedNotification(view: DataView): number | null {
+        }
+
+        // GPD SrcID field
+        if (frameType === ZgpConstants.GpNwkDataFrame && applicationId === 0) {
+            srcId = buf.readUInt32();
+        } else if (
+            frameType === ZgpConstants.GpNwkMaintenanceFrame &&
+            nwkFrameControl & ZgpConstants.GpNwkFrameControlExtensionFlag &&
+            applicationId === 0
+        ) {
+            srcId = buf.readUInt32();
```
Should combine in one `if`, same inner code.

Nerivec commented on src/adapter/deconz/driver/frameParser.ts:
```diff
@@ -440,41 +441,152 @@ function parseDeviceStateChangedNotification(view: DataView): number | null {
+        const applicationId = nwkExtFrameControl & 7;
+        const securityLevel = (nwkExtFrameControl >> 3) & 3;
+
+        if (applicationId > 2) {
```
Might want to put a few `const enum` (stripped out at compile time) at the top for some of these to avoid magic numbers.
_Feel free to copy from https://github.com/Nerivec/zigbee-on-host/blob/e6a56389bb51002ee32a77ab7117851e4b85475a/src/zigbee/zigbee-nwkgp.ts#L30-L59 ._

Nerivec commented on src/adapter/deconz/driver/frameParser.ts:
```diff
@@ -440,41 +441,152 @@ function parseDeviceStateChangedNotification(view: DataView): number | null {
+        let frameCounter = 0;
+        let commandId = 0;
+        let commandFrameSize = 0;
+        let commandFrame = null;
```
```ts
let commandFrame: Buffer | undefined;
```
more explicit typing wouldn't hurt here I think.

manup commented on src/adapter/deconz/driver/frameParser.ts:
```diff
@@ -440,41 +441,152 @@ function parseDeviceStateChangedNotification(view: DataView): number | null {
+        let commandFrameSize = 0;
+        let commandFrame = null;
+
+        const buf = new Buffalo(Buffer.from(view.buffer));
```
At least for now, the incoming packet handler currently provides a `DataView` I'd like to refactor this in a upcoming PR to use plain `Buffer` and `Buffalo` for all parser functions. The DataView is used since we didn't understand the Buffalo interface earlier.

Nerivec replied:
Ah right, the `.buffer` is an ArrayBuffer, forgot about that!
Sounds good about the refactor.

manup commented:
@Nerivec thanks a lot for the review :+1: The updated PR hopefully addresses all comments expect the initial buffer copy which I'd like to address in another PR to get rid of `DataView`.

Nerivec reviewed (approved):
👍 

Note: `enum` != `const enum` (only the latter is inlined at compile time) https://www.typescriptlang.org/docs/handbook/enums.html#const-enums
